Sites Dashboard: Sync checklist progress with the dashboard checklist

vykes-mac opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Closes https://github.com/Automattic/wp-calypso/issues/100260

## Proposed Changes

- Update `useMyHomeCardLaunchpad` hook to support optional site ID
- Replace custom SVG donut progress with CircularProgressBar component
- Improve site launch nag rendering with dynamic progress tracking
- Adjust site nag height so that all rows on the dashboard have the same height
  - Prevents layout shift as site nags are dynamically loaded
- Ensure donut is sync'd, regardless of whether the site has the focused launchpad or not

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* As part of code blue issues to improve platform overall consistency 

The site nag and status label have less padding between each other now. This is how the rows are able to be all the same height. It does look a bit more cramped. But on the other hand, the site title and site URL are also sort of cramped, and this allows the text to line up better, like so:

![CleanShot 2025-03-12 at 21 23 14@2x](https://github.com/user-attachments/assets/cd11842b-d10a-41c9-8bed-3fdaed3d50dc)


## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to `/sites/dashboard`
* Verify the checklist progress is the same as the checklist progress shown in the launchpad
* Clear your cache and reload the dashboard to confirm there's no vertical shift as the data view loads
* Use the dashboard appearance controls to try different "densities" and confirm there's still no layout shift while loading
* Test different sorts of launchpads, e.g.
  * Launchpad of atomic site
    * Create site with `/setup/new-hosted-site`
    * Complete purchase
    * Go through the launchpad tasks on `/home`, comparing with the donut on `/sites` in another tab
  * That has been reverted
    * Create site with `/setup/new-hosted-site`
    * Either bail from the checkout so the site doesn't go atomic, or complete purchase and then revert the site to simple
    * Go through the launchpad tasks on `/home`, comparing with the donut on `/sites` in another tab
  * With focused launchpad
    * Create site with `/setup/onboarding`
    * Choose a sell goal
    * Go through the launchpad tasks, comparing with the donut on `/sites` in another tab
  * With legacy onboarding launchpad
    * Create site with `/start/free`
    * Go through the legacy onboarding screen style launchpad, comparing with the donut on `/sites` in another tab

| Before | After |
|--------|-------|
|  ![image](https://github.com/user-attachments/assets/8990aaac-7171-46c4-a0e5-0a6361d4c345) | ![image](https://github.com/user-attachments/assets/afe66191-ba6d-48c3-a861-fdaf50f51021)


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

vykes-mac committed:
Sites Dashboard: Enhance Launchpad Card with Progress Tracking

- Update `useMyHomeCardLaunchpad` hook to support optional site ID
- Add loading state and site ID support to Launchpad card
- Replace custom SVG donut progress with CircularProgressBar component
- Improve site launch nag rendering with dynamic progress tracking

p-jackson commented on client/sites-dashboard/components/sites-site-launch-nag.tsx:
```diff
@@ -1,37 +1,30 @@
 import styled from '@emotion/styled';
 import { useI18n } from '@wordpress/react-i18n';
 import { useInView } from 'react-intersection-observer';
+import { useMyHomeCardLaunchpad } from 'calypso/my-sites/customer-home/cards/launchpad/use-my-home-card-launchpad';
```
I'm thinking we should import `useLaunchpad` from `@automattic/data-stores` instead. It seems better to have the dashboard depend on a package rather than depend on Calypso code.

The `useLaunchpad` hook also already includes a site parameter.

p-jackson commented on client/sites-dashboard/components/sites-site-launch-nag.tsx:
```diff
@@ -106,7 +92,8 @@ export const SiteLaunchNag = ( { site }: SiteLaunchNagProps ) => {
 			} }
 		>
-			<SiteLaunchDonut /> <SiteLaunchNagText>{ text }</SiteLaunchNagText>
+			<SiteLaunchDonut numberOfSteps={ numberOfSteps } completedSteps={ completedSteps } />
```
It's a little more complicated than just using `numberOfSteps` and `completedSteps`. The focused launchpad and card version of the launchpad use different math.

The focused launchpad actually subtracts launch tasks from the counts.

https://github.com/Automattic/wp-calypso/blob/3447e403c5d8f8749c882b0fe8c338720e05f4b8/client/my-sites/customer-home/components/full-screen-launchpad.tsx#L130-L131

Which could lead to subtly different donuts. E.g. my focused launchpad is at 5/5 but my dashboard donut is at 5/6

![CleanShot 2025-03-10 at 12 04 17@2x](https://github.com/user-attachments/assets/ab92f241-cf56-4db4-bd88-89a56210e86f)

![CleanShot 2025-03-10 at 12 05 13@2x](https://github.com/user-attachments/assets/32557f0b-0e10-48c7-bb09-fd0f668fb0a6)

p-jackson replied:
Actually I've asked about this case https://github.com/Automattic/wp-calypso/issues/100260#issuecomment-2709129315

p-jackson reviewed (commented):
The donut icon on the dashboard actually makes the row height a bit larger. It's always done this, but now that the icon loads async, it causes a noticeable layout shift.


https://github.com/user-attachments/assets/70ea4ea3-c90d-45a1-b92a-a5b143f67d97


Overall the amount of things that load async on the dashboard does make it feel a bit chaotic IMO. But I think since we're adding one more thing that loads async the least we could do is not make things worse by introducing the layout shift.

p-jackson committed:
Site dashboard uses useLaunchpad from @automattic/data-stores

p-jackson committed:
The SiteLaunchDonut doesn't take up any more space than the text

This means rows that show a SiteLaunchDonut are no taller than those
that don't. This prevents layout shift as the dotnut loads

p-jackson committed:
Site dashboard donut matches the focused launchpad donut

p-jackson committed:
Request launchpad using site slug so that we can make use of cache

p-jackson committed:
"Hosted sites" that have been reverted show the legacy checklist

p-jackson commented:
Pinging @fcoveram just a heads up about this change I made to the padding in the site status column.

![image](https://github.com/user-attachments/assets/b31f34a3-f460-44b3-860b-afa9d66ac922)

It feels a bit tight, but it also means that all the rows now have the same height and the text baselines of these two columns are now mostly aligned.

autumnfjeld commented:
✅  Verify the checklist progress is the same as the checklist progress shown in the launchpad

<img width="570" alt="image" src="https://github.com/user-attachments/assets/42c08195-fbf3-4679-8afd-bbf273a6efe6" />

<img width="844" alt="image" src="https://github.com/user-attachments/assets/af487f52-2c23-4e07-a778-d3042a1db096" />


---------

<img width="565" alt="image" src="https://github.com/user-attachments/assets/33368247-047a-4fea-8759-f13ee0251870" />


<img width="847" alt="image" src="https://github.com/user-attachments/assets/59447f63-211c-47c2-ba82-affd0cda07c6" />

------
<img width="561" alt="image" src="https://github.com/user-attachments/assets/19be6192-c448-4006-8953-38112a199c70" />

<img width="840" alt="image" src="https://github.com/user-attachments/assets/bed0204b-7fb4-4fcd-80c6-c1e607f01071" />

autumnfjeld commented:
✅  Clear your cache and reload the dashboard to confirm there's no vertical shift as the data view loads.

Very smooth. Nice.

autumnfjeld commented:
✅ 
> Launchpad of atomic site
> Create site with /setup/new-hosted-site
> Complete purchase
> Go through the launchpad tasks on /home, comparing with the donut on /sites in another tab

Tested on a non-a8c account

<img width="686" alt="image" src="https://github.com/user-attachments/assets/ad162d05-0738-4b92-87d9-1d2114a1db4a" />

<img width="971" alt="image" src="https://github.com/user-attachments/assets/d2c5894f-5194-41c7-84ad-3070bd976db8" />

autumnfjeld commented:
✅ 
> With focused launchpad
> Create site with /setup/onboarding
> Choose a sell goal
> Go through the launchpad tasks, comparing with the donut on /sites in another tab


<img width="722" alt="image" src="https://github.com/user-attachments/assets/096504e6-7e4e-4e61-b2dd-f1130e862119" />


<img width="1117" alt="image" src="https://github.com/user-attachments/assets/a525bf75-304e-41fd-bfe6-43fcd2d412a0" />

autumnfjeld commented:
✅ 
> With legacy onboarding launchpad
> Create site with /start/free
> Go through the legacy onboarding screen style launchpad, comparing with the donut on /sites in another tab

<img width="698" alt="image" src="https://github.com/user-attachments/assets/eb65bf0f-35da-405c-8914-e7dd1d5daf37" />

<img width="1460" alt="image" src="https://github.com/user-attachments/assets/1f290cf7-5904-4605-a3ce-0458cf7110a8" />



(At this point in my testing I'm ready to make a call that we remove all  variations of Launchpads!!!!  😂)

autumnfjeld commented:
✅ 
> That has been reverted
> Create site with /setup/new-hosted-site
> Either bail from the checkout so the site doesn't go atomic, or complete purchase and then revert the site to simple
> Go through the launchpad tasks on /home, comparing with the donut on /sites in another tab

<img width="707" alt="image" src="https://github.com/user-attachments/assets/00562b68-3b1b-4971-8790-17e2340154a4" />

<img width="975" alt="image" src="https://github.com/user-attachments/assets/b04b5598-c25c-4251-81d0-8bbcd6dab085" />

vykes-mac commented:
Thanks @p-jackson for holding the baton on this one! ✨
